### ISSUE TITLE
Unable to manually replicate DTFT

### ISSUE THREAD

**Issue by Arjun-Khurana** [NanoComp/meep#2183]
I have a simple ring resonator structure (based on [this tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#modes-of-a-ring-resonator)). 

I place a dft fields object in the ring using `add_dft_fields` with argument `freqs` set to the center of the Gaussian excitation, and at the same location, I call `get_field_point` every 0.5 Meep time steps. After running the simulation, I manually calculate the dtft for the center frequency of the Gaussian pulse using the following formula:

$$\hat{f}(w) = \frac{\Delta t}{\sqrt{2\pi}}\sum_n{f(n \Delta t) e^{i 2 \pi f_{cen} n \Delta t }}$$

Where `fcen` is the center frequency of the Gaussian excitation pulse. 

I compare the result of my manual dtft calculation with the built in calculation:

```
abs(meep): 0.47794924
abs(manual): 8.380444581645783

ang(meep): 1.09795605
ang(manual): 0.20224952527716047
```

The results clearly do not match. I assume I am doing something elementary incorrectly, but am unsure what it is. 

Full code is [here](https://github.com/Arjun-Khurana/series-extrapolation).

**Comment by smartalecH** [comment]
@Arjun-Khurana can you post your latest results?

**Comment by stevengj** [comment]
To do something on *every* Meep timestep, I would just call `mp.at_every` with `dt=0` (we should document this).     (Note that the first call should then be at timestep 0, so the times in your DTFT will then be `sim.fields.dt * [0,1,2,...]`.)

You can then also extract the Meep Δt (after the run completes) with `sim.fields.dt`.

**Comment by stevengj** [comment]
Another thing you could check would be to add a `printf` to output the fields that Meep is using internally for its DFT:  https://github.com/NanoComp/meep/blob/master/src/dft.cpp#L298-L299

(Note that this includes an interpolation weight factor `w` that is later summed over grid points as well as a `scale` factor that includes the  Δt/sqrt(2π).)

**Comment by Arjun-Khurana** [comment]
> @Arjun-Khurana can you post your latest results?

Using `dt = 0` and multiple frequency points instead of a single center frequency, I can now reasonably replicate the DTFT, with one caveat: I have to include a fudge factor of 1/4 in my dtft calculation. @smartalecH hypothesized that this may be something to do with how the sources are implemented. 

New formula:

$$\hat{f}(\omega)=\frac{\Delta t}{4 \sqrt{2 \pi}} \sum_n{f(n \Delta t)e^{i\omega n \Delta t}}$$

![abs_dtft](https://user-images.githubusercontent.com/35874188/183124470-0c4bfc1d-ff47-43ba-9e0c-593342be85c0.png)

![angle_dtft](https://user-images.githubusercontent.com/35874188/183124484-20f8962a-d7de-4cea-9405-735a9258864e.png)

**Comment by MarkMa1990** [comment]
> > @Arjun-Khurana can you post your latest results?
> 
> Using `dt = 0` and multiple frequency points instead of a single center frequency, I can now reasonably replicate the DTFT, with one caveat: I have to include a fudge factor of 1/4 in my dtft calculation. @smartalecH that this may be something to do with how the sources are implemented.
> 
> New formula:
> 
> f^(ω)=Δt42π∑nf(nΔt)eiωnΔt
> 
> The updated code is in my [series-extrapolation repository](https://github.com/Arjun-Khurana/series-extrapolation)
> 
> ![abs_dtft](https://user-images.githubusercontent.com/35874188/183124470-0c4bfc1d-ff47-43ba-9e0c-593342be85c0.png)
> 
> ![angle_dtft](https://user-images.githubusercontent.com/35874188/183124484-20f8962a-d7de-4cea-9405-735a9258864e.png)

is it complex field simulation or real field?

**Comment by Arjun-Khurana** [comment]
> > > @Arjun-Khurana can you post your latest results?
> > 
> > 
> > Using `dt = 0` and multiple frequency points instead of a single center frequency, I can now reasonably replicate the DTFT, with one caveat: I have to include a fudge factor of 1/4 in my dtft calculation. @smartalecH that this may be something to do with how the sources are implemented.
> > New formula:
> > f^(ω)=Δt42π∑nf(nΔt)eiωnΔt
> > The updated code is in my [series-extrapolation repository](https://github.com/Arjun-Khurana/series-extrapolation)
> > ![abs_dtft](https://user-images.githubusercontent.com/35874188/183124470-0c4bfc1d-ff47-43ba-9e0c-593342be85c0.png)
> > ![angle_dtft](https://user-images.githubusercontent.com/35874188/183124484-20f8962a-d7de-4cea-9405-735a9258864e.png)
> 
> is it complex field simulation or real field?

The simulation has real fields, the DTFT is complex valued.

**Comment by MarkMa1990** [comment]
> > > > @Arjun-Khurana can you post your latest results?
> > > 
> > > 
> > > Using `dt = 0` and multiple frequency points instead of a single center frequency, I can now reasonably replicate the DTFT, with one caveat: I have to include a fudge factor of 1/4 in my dtft calculation. @smartalecH that this may be something to do with how the sources are implemented.
> > > New formula:
> > > f^(ω)=Δt42π∑nf(nΔt)eiωnΔt
> > > The updated code is in my [series-extrapolation repository](https://github.com/Arjun-Khurana/series-extrapolation)
> > > ![abs_dtft](https://user-images.githubusercontent.com/35874188/183124470-0c4bfc1d-ff47-43ba-9e0c-593342be85c0.png)
> > > ![angle_dtft](https://user-images.githubusercontent.com/35874188/183124484-20f8962a-d7de-4cea-9405-735a9258864e.png)
> > 
> > 
> > is it complex field simulation or real field?
> 
> The simulation has real fields, the DTFT is complex valued.

yes, but I mean is the simulation in complex fields or just real? in meep, as soon as k_point is set to mp.Vector3(x,x,x), it automatically assumed a bloch boundary condition, and if the three components are all non-zero, then the simulation is complex fields. I am wondering the 1/4 factor comes from the complex fields or real field simulations.

**Comment by stevengj** [comment]
To track down the 1/4 factor, maybe add a `printf` into the `update_dfts` method to see what fields it's DTFT-ing?
